File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.ProtectionProxyExample/SensitiveDocument.cs
namespace ProxyDesignPattern.ProtectionProxyExample;

/// <summary>
/// RealSubject - The actual sensitive document.
/// Contains the real implementation without any access control.
/// </summary>
public class SensitiveDocument : IDocument
{
    private readonly string _name;
    private string _content;

    public SensitiveDocument(string name, string content)
    {
        _name = name;
        _content = content;
    }

    public void Read()
    {
        Console.WriteLine($"[Document] Reading '{_name}':");
        Console.WriteLine($"[Document] Content: {_content}");
    }

    public void Write(string content)
    {
        _content = content;
        Console.WriteLine($"[Document] Content updated to: {_content}");
    }

    public void Delete()
    {
        Console.WriteLine($"[Document] '{_name}' has been deleted!");
        _content = string.Empty;
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.ProtectionProxyExample/src/SensitiveDocument.ts
import { IDocument } from "./IDocument";

/**
 * RealSubject - The actual sensitive document.
 * Contains the real implementation without any access control.
 */
export class SensitiveDocument implements IDocument {
  private _content: string;

  constructor(
    private readonly _name: string,
    content: string
  ) {
    this._content = content;
  }

  read(): void {
    console.log(`[Document] Reading '${this._name}':`);
    console.log(`[Document] Content: ${this._content}`);
  }

  write(content: string): void {
    this._content = content;
    console.log(`[Document] Content updated to: ${this._content}`);
  }

  delete(): void {
    console.log(`[Document] '${this._name}' has been deleted!`);
    this._content = "";
  }
}
